src/ifcwrap-csharp/IfcCSharp.i

```diff
@@ -0,0 +1,377 @@
+#endif
+#ifdef HAS_SCHEMA_4x3_rc1
+	#include "../ifcparse/Ifc4x3_rc1.h"
+#endif
+#ifdef HAS_SCHEMA_4x3_rc2
+	#include "../ifcparse/Ifc4x3_rc2.h"
+#endif
+#ifdef HAS_SCHEMA_4x3_rc3
+	#include "../ifcparse/Ifc4x3_rc3.h"
+#endif
+#ifdef HAS_SCHEMA_4x3_rc4
+	#include "../ifcparse/Ifc4x3_rc4.h"
+#endif
+#ifdef HAS_SCHEMA_4x3
+	#include "../ifcparse/Ifc4x3.h"
+#endif
+#ifdef HAS_SCHEMA_4x3_tc1
+	#include "../ifcparse/Ifc4x3_tc1.h"
+#endif
+#ifdef HAS_SCHEMA_4x3_add1
+	#include "../ifcparse/Ifc4x3_add1.h"
+#endif
+#ifdef HAS_SCHEMA_4x3_add2
+	#include "../ifcparse/Ifc4x3_add2.h"
+#endif
+
+	#include "../ifcparse/ArgumentType.h"
+	#include "../ifcparse/Argument.h"
+	#include "../ifcparse/aggregate_of_instance.h"
+	#include "../ifcparse/IfcBaseClass.h"
+	#include "../ifcparse/IfcFile.h"
+	#include "../ifcparse/IfcSchema.h"
+	#include "../ifcparse/utils.h"
+	
+	#include "../ifcgeom/ConversionSettings.h"
+	#include "../ifcgeom/ConversionResult.h"
+
+	#include "../svgfill/src/svgfill.h"
+%}
+
+%include "IfcGeomWrapper.i"
+%include "IfcParseWrapper.i"
+%include "std_vector.i"
+	
+%inline %{
+	
+	template<typename T>
+	struct TypedArgument {
+
+		virtual ~TypedArgument() = default;
+
+		virtual bool HasValue() const = 0;
+		virtual T GetValue() const = 0;
+	};
+
+	template<typename T>
+	struct WithValue : public TypedArgument<T> {
+		explicit WithValue(const T& arg) : value(arg) {}
+		explicit WithValue(const T&& arg) : value(std::move(arg)) {}
+
+		inline bool HasValue() const override {
+			return true;
+		}
+
+		inline T GetValue() const override { 
+			return value;
+		}
+
+		private:
+		T value;
+	};
+
+	template <typename T>
+	struct NoValue : public TypedArgument<T> {
+		inline bool HasValue() const override { return false; }
+		inline T GetValue() const override { 
+			throw std::runtime_error("Has no value");
+		}
+	};
+
+%}
+
+
+%apply const std::string& { std::string* }
+
+namespace std {
+  %template(Vec3) std::array<double, 3>;
+  %template(Vec4) std::array<double, 4>;
+  %template(FloatVector) std::vector<float>;
+  %template(IntVector) std::vector<int>;
+  %template(DoubleVector) std::vector<double>;
+  %template(StringVector) std::vector<std::string>;
+  %template(FloatVectorVector) std::vector<std::vector<float>>;
+  %template(DoubleVectorVector) std::vector<std::vector<double>>;
+
+  %template(MaterialVector) std::vector<std::shared_ptr<ifcopenshell::geometry::taxonomy::style>>;
+
+  %template(ArgumentByType) std::pair<IfcUtil::ArgumentType, Argument*>;
+
+  %template(EntityPtrList) std::vector<IfcUtil::IfcBaseClass*>;
+}
+
+%extend std::array<double, 3> {
+	const double X;
+	const double Y;
+	const double Z;
+}
+
+%{
+	double std_array_Sl_double_Sc_3_Sg__X_get(const std::array<double, 3>* vec) {
+		return (*vec)[0];
+	}
+
+	double std_array_Sl_double_Sc_3_Sg__Y_get(const std::array<double, 3>* vec) {
+		return (*vec)[1];
+	}
+
+	double std_array_Sl_double_Sc_3_Sg__Z_get(const std::array<double, 3>* vec) {
+		return (*vec)[2];
+	}
+%}
```

Comment:
Please add a comment answering the following questions:
- What does it do?
- Why do we need it?
- Where is the swig preprocessor inserting it?
- Why it is called ``std_array_Sl_double_Sc_3_Sg__X_get``